web/src/pages/api/public/media/upload-url.ts

```diff
@@ -0,0 +1,227 @@
+
+        if (contentLength > env.LANGFUSE_S3_MEDIA_MAX_CONTENT_LENGTH)
+          throw new InvalidRequestError(
+            `File size must be less than ${env.LANGFUSE_S3_MEDIA_MAX_CONTENT_LENGTH} bytes`,
+          );
+
+        const { mediaId, uploadUrl } = await prisma.$transaction(async (tx) => {
+          const existingMedia = await tx.media.findUnique({
+            where: {
+              projectId_sha256Hash: {
+                projectId,
+                sha256Hash,
+              },
+            },
+          });
+
+          if (
+            existingMedia &&
+            existingMedia.uploadHttpStatus === 200 &&
+            existingMedia.contentType === contentType
+          ) {
+            if (observationId) {
+              await tx.observationMedia.upsert({
+                where: {
+                  projectId_traceId_observationId_mediaId_field: {
+                    projectId,
+                    traceId,
+                    observationId,
+                    mediaId: existingMedia.id,
+                    field,
+                  },
+                },
+                update: {},
+                create: {
+                  projectId,
+                  traceId,
+                  observationId,
+                  mediaId: existingMedia.id,
+                  field,
+                },
+              });
+            } else {
+              await tx.traceMedia.upsert({
+                where: {
+                  projectId_traceId_mediaId_field: {
+                    projectId,
+                    traceId,
+                    mediaId: existingMedia.id,
+                    field,
+                  },
+                },
+                update: {},
+                create: {
+                  projectId,
+                  traceId,
+                  field,
+                  mediaId: existingMedia.id,
+                },
+              });
+            }
+
+            return {
+              mediaId: existingMedia.id,
+              uploadUrl: null,
+            };
+          }
+
+          const mediaId = existingMedia?.id ?? randomUUID();
+
+          if (
+            !(
+              env.LANGFUSE_S3_MEDIA_UPLOAD_ENABLED &&
+              env.LANGFUSE_S3_MEDIA_UPLOAD_BUCKET
+            )
+          )
+            throw new InternalServerError(
+              "Media upload to blob storage not enabled or no bucket configured",
+            );
+
+          const s3Client = getMediaStorageServiceClient(
+            env.LANGFUSE_S3_MEDIA_UPLOAD_BUCKET,
+          );
+
+          const bucketPath = getBucketPath({
+            projectId,
+            mediaId,
+            contentType,
+          });
+
+          const uploadUrl = await s3Client.getSignedUploadUrl({
+            path: bucketPath,
+            ttlSeconds: 60 * 60, // 1 hour
+            sha256Hash,
+            contentType,
+            contentLength,
+          });
+
+          await Promise.all([
+            tx.media.upsert({
```

Comment:
cant we go to the database only once in this API?
- first we create the S3 files
- afterwards we insert into the database.